components/wifi/esp8266-driver/ESP8266/ESP8266.cpp

```diff
@@ -635,52 +671,72 @@ nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
         goto END;
     }
 
-    //We might receive "busy s/p..." and "OK" from modem, so we need to check that also
-    _ok_received = false;
-    _parser.oob("OK", callback(this, &ESP8266::_oob_ok_received));
-
     if (!_parser.recv(">")) {
-        _parser.remove_oob("OK");
-        if (_busy) {
-            if (_ok_received) {
-                goto RETRY;
-            } else if (_parser.recv("OK")) {
-                goto RETRY;
-            }
-        }
+        // This means ESP8266 hasn't even started to receive data
         tr_debug("send(): Didn't get \">\"");
-        ret = NSAPI_ERROR_WOULD_BLOCK;
+        if (_sock_i[id].proto == NSAPI_TCP) {
+            ret = NSAPI_ERROR_WOULD_BLOCK; // Not necessarily critical error.
+        } else if (_sock_i[id].proto == NSAPI_UDP) {
+            ret = NSAPI_ERROR_NO_MEMORY;
+        }
+        goto END;
+    }
+
+    if (_parser.write((char *)data, (int)amount) < 0) {
+        tr_debug("send(): Failed to write serial data");
+        // Serial is not working, serious error, reset needed.
+        ret = NSAPI_ERROR_DEVICE_ERROR;
         goto END;
     }
-    _ok_received = false;
-    _parser.remove_oob("OK");
 
-    if (_parser.write((char *)data, (int)amount) >= 0 && _parser.recv("SEND OK")) {
-        ret = NSAPI_ERROR_OK;
+    // The "Recv X bytes" is not documented.
+    if (!_parser.recv("Recv %d bytes", &bytes_confirmed)) {
+        tr_debug("send(): Bytes not confirmed.");
+        if (_sock_i[id].proto == NSAPI_TCP) {
+            ret = NSAPI_ERROR_WOULD_BLOCK;
+        } else if (_sock_i[id].proto == NSAPI_UDP) {
+            ret = NSAPI_ERROR_NO_MEMORY;
+        }
+        goto END;
+    } else if (bytes_confirmed != amount) {
+        tr_debug("send(): Error: confirmed %d bytes, but expected %d.", bytes_confirmed, amount);
+        ret = NSAPI_ERROR_DEVICE_ERROR;
```

Comment:
Would it be ok in case of TCP that ESP8266 accepts less data than we are trying to send?

Reply:
I thought about this as well, but wasn't brave enough to implement it ;-). I have never seen this situation to happen, so I assume it's either all or nothing, but you are right, if ESP ever decides to accept partial write, we should be able to handle it.